hudi-client/src/main/java/org/apache/hudi/metadata/HoodieMetadataImpl.java

```diff
@@ -0,0 +1,1003 @@
+    metaClient = HoodieTableMetaClient.initTableType(hadoopConf, metadataBasePath.toString(),
+        HoodieTableType.MERGE_ON_READ, tableName, "archived", HoodieMetadataPayload.class.getName(),
+        HoodieFileFormat.HFILE.toString());
+
+    // List all partitions in the basePath of the containing dataset
+    FileSystem fs = FSUtils.getFs(basePath, hadoopConf);
+    List<String> partitions = FSUtils.getAllPartitionPaths(fs, basePath, false);
+    LOG.info("Initializing metadata table by using file listings in " + partitions.size() + " partitions");
+
+    // List all partitions in parallel and collect the files in them
+    final String dbasePath = basePath;
+    final SerializableConfiguration serializedConf = new SerializableConfiguration(hadoopConf);
+    int parallelism =  Math.min(partitions.size(), 100) + 1; // +1 to ensure non zero
+    JavaPairRDD<String, FileStatus[]> partitionFileListRDD = jsc.parallelize(partitions, parallelism)
+        .mapToPair(partition -> {
+          FileSystem fsys = FSUtils.getFs(dbasePath, serializedConf.get());
+          FileStatus[] statuses = FSUtils.getAllDataFilesInPartition(fsys, new Path(dbasePath, partition));
+          return new Tuple2<>(partition, statuses);
+        });
+
+    // Collect the list of partitions and file lists
+    List<Tuple2<String, FileStatus[]>> partitionFileList = partitionFileListRDD.collect();
+
+    // Create a HoodieCommitMetadata with writeStats for all discovered files
+    int[] stats = {0};
+    HoodieCommitMetadata metadata = new HoodieCommitMetadata();
+    partitionFileList.forEach(t -> {
+      final String partition = t._1;
+      try {
+        if (!fs.exists(new Path(basePath, partition + Path.SEPARATOR + HoodiePartitionMetadata.HOODIE_PARTITION_METAFILE))) {
+          return;
+        }
+      } catch (IOException e) {
+        throw new HoodieMetadataException("Failed to check partition " + partition, e);
+      }
+
+      // If the partition has no files then create a writeStat with no file path
+      if (t._2.length == 0) {
+        HoodieWriteStat writeStat = new HoodieWriteStat();
+        writeStat.setPartitionPath(partition);
+        metadata.addWriteStat(partition, writeStat);
+      } else {
+        Arrays.stream(t._2).forEach(status -> {
+          String filename = status.getPath().getName();
+          if (filename.equals(HoodiePartitionMetadata.HOODIE_PARTITION_METAFILE)) {
+            return;
+          }
+          HoodieWriteStat writeStat = new HoodieWriteStat();
+          writeStat.setPath(partition + Path.SEPARATOR + filename);
+          writeStat.setPartitionPath(partition);
+          writeStat.setTotalWriteBytes(status.getLen());
+          metadata.addWriteStat(partition, writeStat);
+        });
+      }
+      stats[0] += t._2.length;
+    });
+
+    LOG.info("Committing " + partitionFileList.size() + " partitions and " + stats[0] + " files to metadata");
```

Comment:
if we fail doing this i.e the initial commit with all the existing files. the metadata table will be created, but no commits. how do we detect this and attempt reiniting next time around?

Reply:
`    boolean exists = fs.exists(new Path(metadataBasePath));` , I am saying exists will be equal to true, even if we fail to do the initing and fail midway?